cpp/src/utility/scatter_output_iterator.cuh

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <thrust/distance.h>
+#include <thrust/iterator/iterator_adaptor.h>
+
+#pragma once
+
+namespace cuspatial {
+namespace detail {
+
+/** @brief scatters outputs to a given index, according to the input scatter map.
+ *
+ * @note: if the scatter map produces a negative index, the corrosponding output is ignored.
+ **/
+template <typename OutputIterator, typename ScatterIterator>
+class scatter_output_iterator_proxy {
+ public:
+  __host__ __device__ scatter_output_iterator_proxy(const OutputIterator& begin,
+                                                    const OutputIterator& out,
+                                                    const ScatterIterator& scatter_map)
+    : begin(begin), out(out), scatter_map(scatter_map)
+  {
+  }
+
+  template <typename T>
+  __host__ __device__ scatter_output_iterator_proxy operator=(const T& element)
+  {
+    auto scatter_idx = *(scatter_map + thrust::distance(begin, out));
+
+    if (scatter_idx >= 0) { *(begin + scatter_idx) = element; }
+
+    return *this;
+  }
+
+ private:
+  OutputIterator begin;
+  OutputIterator out;
+  ScatterIterator scatter_map;
+};
+
+template <typename OutputIterator, typename ScatterIterator>
+class scatter_output_iterator;
```

Comment:
Wow there are lots of classes, structs, proxies. Please document, I'm confused.

Reply:
Attempted to address this concern in latest commits, but ultimately it's the same iterator proxy pattern used by Thrust. At first I didn't understand it either. :man_shrugging: